Project2015To2017/Reading/ProjectReader.cs

```diff
@@ -1,28 +1,48 @@
+using Project2015To2017.Definition;
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.IO;
 using System.Linq;
 using System.Xml;
 using System.Xml.Linq;
-using Project2015To2017.Definition;
 using static Project2015To2017.Definition.Project;
 
 namespace Project2015To2017.Reading
 {
 	public class ProjectReader
 	{
-		public Project Read(string filePath)
+		public static readonly ProjectReader Instance = new ProjectReader();
+		public bool EnableCaching { get; set; }
+
+		public Project Read(string filePath, IProgress<string> progress = null)
 		{
-			return Read(filePath, new Progress<string>(_ => { }));
+			progress = progress ?? new Progress<string>(_ => { });
+			return Read(new FileInfo(filePath), progress);
 		}
 
-		public Project Read(string filePath, IProgress<string> progress)
+		/// <summary>
+		/// Process-lifetime-long cache of loaded projects
+		/// <remarks>
+		/// It was supposed to be of <see cref="System.Runtime.Caching.ObjectCache"/> type, but that can still remove items from cache under memory pressure,
+		/// while we still need availability guarantees.
+		/// </remarks>
+		/// </summary>
+		private readonly Dictionary<int, Project> _cache = new Dictionary<int, Project>();
+
+		public Project Read(FileInfo file, IProgress<string> progress, bool loadModern = false)
 		{
+			var filePath = file.FullName;
+			var hash = filePath.GetHashCode();
```

Comment:
You can just use the filePath as the key in the Dictionary. It will calculate the hashcode internally anyway as that is how it does fast lookups.

Reply:
Yeah, I know that, but I considered that usually hashtables do store full keys too to verify that there is no collision on adding a new key to the hashtables. I considered that memory efficiency is more important than collision protection and store only hashcode.

Reply:
I would keep the Dictionary to be `Dictionary<FileInfo, Project>` that says more about what you're storing than the hashcode. I think that code readability is more important than perf here, I have seen solutions with ~150 projects but that becomes pretty unwieldy so the memory overhead is minimal here, we're not storing millions of entries.